tools/translate.sh

```diff
@@ -0,0 +1,120 @@
+#!/bin/sh
+# Shell script helper for translating Metamath tool files to bash
+# Mario Carneiro 7-Jan-2022
+
+usage() {
+  cat << HELP
+TRANSLATE - Translate a TOOLS script into a corresponding bash script
+
+This command will take a script file intended for use with metamath
+tools-mode, and translate it into a bash script with the same behavior.
+See also SUBMIT. The script is passed the location of the tools
+directory (containing the core script files), which is also the location
+of this file. It uses 'mm-tools' by default, so if you symlink 'mm-tools'
+to the tools directory then these scripts will work.
+
+If the -r (recursive) option is passed, the generated file will translate
+SUBMIT calls to also make use of the same TOOLS directory; otherwise
+it will leave them off (using the default "mm-tools" directory).
+
+If the -e (echo) option is passed, the generated file will contain an
+echo statement for each executed command.
+
+Syntax:  translate.sh [-r] [-e] [TOOLS] < INFILE > OUTFILE
+HELP
+}
+
+if [ $# -eq 1 ] && [ "$1" = "-h" -o "$1" = "--help" ]; then usage; exit; fi
+
+if [ "$1" = "-r" ]; then recursive=1; shift; fi
+if [ "$1" = "-e" ]; then echo=1; shift; fi
+
+if [ $# -eq 0 ]; then tools="mm-tools"
+elif [ $# -eq 1 ]; then tools=$1
+else usage; exit 1; fi
+
+echo "#!/bin/bash"
+awk -v tools="$tools" -v echo="$echo" -v recursive="$recursive" \
+'BEGIN {
+  if (recursive) {
+    extraArg = tools;
+    gsub(/[\\$"`]/, "\\\\&", extraArg);
+    extraArg = " \"" extraArg "\"";
+  }
+  simpleCommands["ADD"] = 1;
+  simpleCommands["DELETE"] = 1;
+  simpleCommands["SUBSTITUTE"] = 1;
+  simpleCommands["SWAP"] = 1;
+  simpleCommands["INSERT"] = 1;
+  simpleCommands["BREAK"] = 1;
+  simpleCommands["BUILD"] = 1;
+  simpleCommands["MATCH"] = 1;
+  simpleCommands["SORT"] = 1;
+  simpleCommands["UNDUPLICATE"] = 1;
+  simpleCommands["DUPLICATE"] = 1;
+  simpleCommands["UNIQUE"] = 1;
+  simpleCommands["REVERSE"] = 1;
+  simpleCommands["RIGHT"] = 1;
+  simpleCommands["PARALLEL"] = 1;
+  simpleCommands["TAG"] = 1;
+} {
+  if ($0 == "") print;
+  else if (substr($0, 1, 1) == "!")
+    print "#" substr($0, 2);
+  else if (toupper($1) == "TOOLS")
+    ;
+  else {
+    $1 = toupper($1);
+    if ($1 == "C") $1 = "COPY"; else
+    if ($1 == "B") $1 = "BEEP"; else
+    if ($1 == "S") $1 = "SUBSTITUTE"; else
+    if ($1 == "T") $1 = "TAG"; else
+    if ($1 == "EXIT" || $1 == "QUIT") exit;
+    if (echo) {
+      echoLine = $0;
+      gsub(/[\\$"]/, "\\\\&", echoLine);
+      print "echo \"" echoLine "\"";
+    }
+    file = tools "/" tolower($1) ".sh";
+    if (simpleCommands[$1]) {
+      infile = $2;
+      $2 = file;
+      $1 = tools "/inplace.sh " infile;
```

Comment:
The idea of a simple command is that it follows the syntax rule
```<cmd> <file> <other args>```
in metamath scripts.
This is an implicit assumption here.  Make it explicit by adding a comment to this respect in line 79 or 44

Reply:
It's not an assumption, it's a classification of the scripts. It just so happens that most of the TOOLS commands have this form, so they can be handled in one go. It's not a user facing classification at all, and it is explicitly enumerated so there are no implicit assumptions here. Maybe the name could be better, but it doesn't seem very important.

Reply:
I tested this block of code, naively entering the parameters in the order required by the script usage help
echo 'add x y dummy-file-name' | ./translate.sh path-to-tools
and of course it failed.

What is the reason behind this failure? I never wrote a metamath script, so I was completely unaware of the grammar details of this language. My assumption was, it parallels that of the bash scripts.  Which is not exactly true.

In the same manner this program (or chunk of code) assumes a particular format of the input script.  This is part of the contract (Design by contract https://en.wikipedia.org/wiki/Design_by_contract) between the user of this script and you as a programmer.  As with real life contracts, their clauses should be noted somewhere.  And somewhere does not mean "elsewhere", but either here, or at least referenced here.

This contract thinking is the result of ages of sloppy programming.  It counteracts situations like that in the beginning of this post.  And tedious debugging sessions.

My two cents

Reply:
I'm confused. The documentation of `translate.sh` says that it is meant to translate TOOLS scripts, so it uses the syntax for tools scripts (and indeed it can't do anything else). The whole point of it is to translate between the metamath argument order and syntax conventions to bash syntax and conventions. If you want to use the shell files directly then just do that can't don't call `translate.sh`.

If you want to know what the grammar details of the metamath TOOLS language are, you should consult the metamath TOOLS help. I suppose I could reproduce the entirety of that documentation inside the help of `translate.sh`, but the main purpose of this command is for "modernizing your old scripts" so the presumption is that you already know the TOOLS syntax or else you have a file written previously with knowledge of this syntax and are interested in converting it.

You are preaching to the choir re: design by contract, but metamath.exe is currently in a shambles with respect to this criterion and it will take a while to get to that point. The purpose of these scripts is to get TOOLS out of metamath.exe and replace them with something good enough. They are *not* air tight regarding error handling for a bunch of reasons, they *will* break if you push on them too hard. The documentation on the scripts is passable but not comprehensive. (If you want to make a documentation suggestion, just post it.) And they mostly either do things that already exist as unix commands or do a weird variation that is not generally useful. In short, I wouldn't recommend anyone use them for new scripts, they exist only for compatibility purposes.

PS: I mentioned earlier that metamath uses low ascii without acknowledging its limitations. Just knowing that this is how the program works makes it easy to trigger bugs:

```bash
$ echo -e '\x03' > test.mm
$  ./metamath test.mm
Metamath - Version 0.199.pre 7-Aug-2021       Type HELP for help, EXIT to exit.
MM> READ "test.mm"
Reading source file "test.mm"... 2 bytes

?Error on line 1 of file "test.mm":
?BUG CHECK:  *** DETECTED BUG 1514
...
```

Reply:
Let me first illustrate how I had written the code:
```
if (stdformattedCmds[$1]) {
      # $0 pre (mm): (mm) <command> <file modified> optional <other args>
      # $0 post (bash): toolspath/inplace.sh <file modified> <command>  <other args>
      inoutfile = $2;
      $2 = cmdAsBash;
      $1 = toolspath "/inplace.sh " inoutfile;
}
```
1. You can easily verify the code does what it is expected to do.  The transformation is complete, no post-processsing necessary, nothing to be shared with other cases.
2. The contract of this piece of code (not that of the whole program) is explicitly given.  Given this is done with the other cases, too, one can verify the divide in fact sums up to the conquer (a common post-condition).
3. The variable names reflect their semantics (more) correctly.

I do not expect this exactly in this form, this has a bit of a showcase already.  But (to my liking) the support for a casual reader is too sparse in the current form.  That my premature interpretation lead to the above mentioned failure is an indication of that.  Of course it was still my fault, I should have been more careful; this could have spared me an extra round in testing, avoiding scratching my head while guessing at what went wrong.

Couldn't I have looked simply at the code, and read from it what is going on?  It's just four lines...  No.  I am not confident, no error crept in, particularly not when reviewing.

This leads to the general question, what quality standard shall we follow in future.  From the alleged scrutiny applied to the Webb telescope down to "works for me most of the time" is a wide range.  I am not in the position to set the standard here.  We shall definitely talk about how to improve review sessions, what was good/bad (https://google.github.io/eng-practices/review/reviewer/standard.html).  For now I follow the path I am used in my company, where I occasionally review other people's code.  Actually I am less rigorous here.  This is due to my lack of familiarity with the code base in total, and the techniques applied in the sources I read.

Reply:
Okay. Personally I'd like to move on from this PR, so feel free to make any changes you like. This is throw-away code, so I don't think a high quality standard is necessitated in this instance. ("Works on `min.cmd` and the 3 or 4 other existing tools scripts is good enough for me.) I'm much more concerned with getting the C code up to scratch, and right now I'm bottlenecked on review. (#36 should be easy, and #37 is big but I did my best to make it straightforward.)